tests/ws_async_shutdown.c

```diff
@@ -0,0 +1,106 @@
+/*
+    Copyright (c) 2012 Martin Sustrik  All rights reserved.
+    Copyright (c) 2015-2016 Jack R. Dunaway. All rights reserved.
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"),
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom
+    the Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included
+    in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+*/
+
+#include "../src/nn.h"
+#include "../src/pubsub.h"
+
+#include "testutil.h"
+#include "../src/utils/thread.c"
+
+/*  Test condition of closing sockets that are blocking in another thread. */
+
+#define SOCKET_ADDRESS "ws://127.0.0.1:5558"
+#define TEST_LOOPS 10
+#define TEST_THREADS 10
+
+static void routine (NN_UNUSED void *arg)
+{
+    int s;
+    int rc;
+    char msg[1];
+
+    nn_assert (arg);
+
+    s = *(int *)arg;
+
+    while (1) {
+        rc = nn_recv (s, &msg, sizeof(msg), 0);
+        if (rc == 0) {
+            continue;
+        }
+
+        nn_assert (rc == -1);
+
+        /*  A timeout is OK since PUB/SUB is lossy. */
+        if (nn_errno () == ETIMEDOUT) {
+            continue;
+        }
+        break;
+    }
+    /*  Socket is expected to be closed by caller.  */
+    errno_assert (nn_errno () == EBADF);
+}
+
+int main ()
```

Comment:
Even further, this test needs the full-blown upgrade made by @Snaipe to dynamically retrieve its port. (This test predates that upgrade.) Stay tuned ...